tf_agents/networks/multi_categorical_projection_network.py

```diff
@@ -0,0 +1,140 @@
+    """Creates an instance of MultiCategoricalProjectionNetwork
+
+    Args:
+      sample_spec: A collection of `tensor_spec.BoundedTensorSpec` detailing
+        the shape and dtypes of samples pulled from the output distribution.
+      logits_init_output_factor: Output factor for initializing kernal
+        logits weights.
+      name: A string representing the name of the network.
+    """
+
+    categories_shape = self._categories_shape(sample_spec)
+    n_unique_categories = np.sum(categories_shape)
+    output_spec = self._output_distribution_spec(
+      [n_unique_categories],
+      sample_spec,
+      categories_shape,
+      name)
+
+    super(MultiCategoricalProjectionNetwork, self).__init__(
+      input_tensor_spec=None,
+      state_spec = (),
+      output_spec=output_spec,
+      name=name
+    )
+
+    self._sample_spec = sample_spec
+    self.categories_shape = categories_shape
+    self.n_unique_categories = n_unique_categories
+
+    self._projection_layer = tf.keras.layers.Dense(
+        self.n_unique_categories,
+        kernel_initializer=tf.compat.v1.keras.initializers.VarianceScaling(
+            scale=logits_init_output_factor),
+        bias_initializer=tf.keras.initializers.Zeros(),
+        name='logits'
+        )
+
+  def _categories_shape(self, sample_spec):
+    def _get_n_categories(array_spec):
+      if not tensor_spec.is_bounded(array_spec):
+        raise ValueError(
+          'sample_spec must be bounded. Got: %s.' % type(array_spec))
+
+      if not tensor_spec.is_discrete(array_spec):
+        raise ValueError('sample_spec must be discrete. Got: %s.' % array_spec)
+
+      n_categories = array_spec.maximum - array_spec.minimum +1
+      return n_categories
+
+    flattened_spec = tf.nest.flatten(sample_spec)
+    categories_shape = tf.nest.map_structure(
+      lambda s: _get_n_categories(s),
+      flattened_spec)
+
+    if len(categories_shape) == 1:
+      categories_shape = categories_shape[0]
+
+    return categories_shape
+
+  def _output_distribution_spec(self, output_shape, sample_spec, categories_shape, network_name):
+    input_param_spec = {
+        'logits':
+            tensor_spec.TensorSpec(
+                shape=output_shape,
+                dtype=tf.float32,
+                name=network_name + '_logits'
+            )
+    }
+    return distribution_spec.DistributionSpec(
+        MultiCategoricalDistributionBlock,
+        input_param_spec,
+        sample_spec=sample_spec,
+        categories_shape = categories_shape
+        )
+
+  def call(self, inputs, outer_rank, training=False, mask=None):
+    batch_squash = utils.BatchSquash(outer_rank)
+    inputs = batch_squash.flatten(inputs)
+    inputs = tf.cast(inputs, tf.float32)
+
+    logits = self._projection_layer(inputs, training=training)
+    logits = tf.reshape(logits, [-1] + [self.n_unique_categories])
+    logits = batch_squash.unflatten(logits)
+    if mask is not None:
+      # assume mask is a flattened array for now
```

Comment:
I suppose mask should have the same shape as actions.

Reply:
I'm not sure about the more appropriate approach for this so any advice here would be good; At this stage the outputs of the network exists as logits, i.e. the vector exists as flattened. In this case, `mask` needs to be flattened as well. 

Alternatively, if we want the `mask` to be same shape as actions, we could do the masking during __init__ of `MultiCategoricalDistributionBlock`